spices/SPICE-0004-HTTP-Proxy-Support.adoc

```diff
@@ -0,0 +1,180 @@
+Proxy settings will be added, and this will control how Pkl makes HTTP requests.
+
+== Detailed design
+
+=== Configuring proxies
+
+A proxy can be set via the following, in order of precedence (least to greatest):
+
+1. `~/.pkl/settings` file
+2. `PklProject` file (in `evaluatorSettings`)
+3. CLI Flags/pkl-gradle options/`HttpClient#Builder`/server mode message passing API
+
+==== Environment Variables
+
+Proxies will _not_ be configured via environment variables.
+This maintains consistency with the rest of Pkl; it does not support using environment variables for any other configuration.
+Additionally, while other languages might use environment variables, they have slightly different meaning from one language to another (see link:https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy[]]).
+Not using environment variables avoids a possble problem where the same environment variable influences both Pkl and another runtime, but in different ways.
+
+==== Settings file
+
+A new `proxy` property is added to the global settings file.
+
+[source,pkl]
+----
+module pkl.settings
+
+import "pkl:Proxy"
+
+/// HTTP proxy settings.
+@Since { version = "0.26.0" }
+proxy: Proxy?
+----
+
+==== `PklProject`
+
+A new `proxy` property is added to `EvaluatorSettings` of a PklProject file.
+
+[source,pkl]
+----
+module pkl.Project
+
+import "pkl:Proxy"
+
+class EvaluatorSettings {
+  proxy: Proxy?
+}
+----
+
+[[new-stdlib-module]]
+==== New stdlib module: `pkl.Proxy`
+
+A new stdlib module is added called `pkl.Proxy`, that is used by `pkl.settings` and `pkl.Project`.
+
+[source,pkl]
+----
+/// Settings that control how Pkl talks to HTTP proxies.
+@Since { version = "0.26.0" }
+module pkl.Proxy
+
+/// The proxy to use for HTTP(s) connections.
+///
+/// At the moment, only HTTP proxies are supported.
+///
+/// Example:
+/// ```
+/// local proxyPassword = read("my-proxy-password.txt").text
+/// address = "http://myuser@\(proxyPassword):my.proxy.example.com:5080"
+/// ```
+address: Uri(startsWith("http://"))
+
+/// Hosts to disable connecting to via a proxy.
+///
+/// Values can be either hostnames, or CIDR blocks.
+///
+/// Follows the same rules as cURL's [--noproxy](https://curl.se/docs/manpage.html#--noproxy)
+/// flag.
+noProxy: Listing<String>(isDistinct)
+----
+
+==== CLI flags
+
+The following CLI flags are added to common CLI options:
+
+* `--proxy` -- A proxy URL to connect to
+* `--noproxy` -- A comma-separated list of hosts to not proxy.
+
+==== pkl-gradle
+
+The following options are added to the common options for `PklSpec`:
+
+* `proxy: Property<String>`
+* `noProxy: ListProperty<String>`
+
+==== `HttpClient#Builder`
+
+When Pkl is embedded with in Java, it is preferrable to use the host's proxy settings if possible.
+When Pkl is run standalone, the proxy should be configued via user-facing means.
+
+To support this, `HttpClient#Builder` receives two new builder methods:
+
+* `setProxySelector(java.net.ProxySelector proxySelector)`
```

Comment:
Is it really necessary to expose ProxySelector?

Reply:
I'm open to ideas here!

Reply:
Why did you expose it?

Reply:
The goal is to allow users to configure Pkl's proxy the same way their application's proxy is configured. `setProxy(proxyAddress, List<URI> noProxy)` is our bespoke implementation, and accepting a `java.net.ProxySelector` allows those that are already using it to configure this client the same way.

Some prior art; both `java.net.HttpClient` and okhttp accept `ProxySelector`.

Reply:
Of course java.net.HttpClient accepts ProxySelector. The point is that with every java.net type added, it becomes more difficult to ever have another implementation of org.pkl.util.http.HttpClient (say NettyHttpClient). It’s better to wait until there is a proven need for “those that are already using ProxySelector can configure our client the same way”, then make an informed decision. That’s why I didn’t expose HttpClient.sendAsync—it’s not yet needed, and with virtual threads added in Java 21, it may never be needed.

Reply:
> But with every java.net type added, it becomes more difficult to ever have another implementation of org.pkl.util.http.HttpClient (say NettyHttpClient).

I'm not sure why this is desirable. Can you elaborate?